rust/connlib/snownet/tests/lib.rs

```diff
@@ -79,6 +79,63 @@ fn migrate_connection_to_new_relay() {
     assert_eq!(alice.packets_from(ip("8.8.8.8")).count(), 1);
 }
 
+#[test]
+fn idle_connection_is_closed_after_5_minutes() {
```

Comment:
One common way to hack time for tests is to expose the idle timeout for example in the actual implementation so you can override it in tests.

Any thoughts on doing that here? I think it would be helpful to start testing a lot of this stuff with "sped up" time intervals.

Another option is to use a "time multiplier" that multiples all time constants by some factor that can be changed when testing.

Reply:
I realize this is easier for some languages than others.

Reply:
`snownet` is sans-IO so it doesn't actually rely on any timers internally (except `boringtun` does for a few things). This test here passes in < 1 sec because it advances time on every `progress` step by 100ms without actually waiting for 100ms.

Reply:
> One common way to hack time for tests is to expose the idle timeout for example in the actual implementation so you can override it in tests.

I am not a fan of these kind of "mocks" or "stub" implementations. The way we do this in connlib is by having all functions that need access to the current time take an `Instant` as a parameter! That way, we can plug whatever "current time" in that we want.

Reply:
```
❯ time /home/thomas/src/github.com/firezone/firezone/rust/target/debug/deps/lib-5ac57e722a8b08b5 idle_connection

running 1 test
test idle_connection_is_closed_after_5_minutes ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 6 filtered out; finished in 0.69s

 idle_connection  0.68s user 0.01s system 98% cpu 0.696 total
```

Reply:
> current time take an Instant as a parameter! That way, we can plug whatever "current time" in that we want.

Yeah that is what I had in mind. Ok cool.

Reply:
Note that this test asserts a false positive because boringtun will keep the `idle_timeout` from ever firing because of WireGuard's keepalive.

Reply:
<img width="1147" alt="Screenshot 2024-06-26 at 10 06 23 PM" src="https://github.com/firezone/firezone/assets/167144/0e946bd9-d953-4995-a469-1bc6d8232457">

Reply:
We should be safe removing WireGuard's keepalive, right? _right_? 🥹

Reply:
> Note that this test asserts a false positive because boringtun will keep the `idle_timeout` from ever firing because of WireGuard's keepalive.

I only check on packets that have been sent  / received from connlib. wireguard's keep-alive should not affect this because it is sent internally in `boringtun`.

I have not yet tested this though!

Time to fork boringtun and make it properly sans-IO ... https://github.com/cloudflare/boringtun/pull/393 :clown_face:

Reply:
I would be in support of maintaining a fork. The core WireGuard piece is "finished" so the only changes should be keeping deps up to date and extending it with more control

Reply:
> I only check on packets that have been sent / received from connlib. wireguard's keep-alive should not affect this because it is sent internally in boringtun.

Does that mean WireGuard will emit failed handshakes/keepalives that we log after the idle connection is closed?

Reply:
> I would be in support of maintaining a fork.

I'd probably even vendor it to be honest. Maintaining a fork is too much effort with additional CI, dependabot etc.

Reply:
I don't think WireGuard's keepalive is doing anything if we're keeping the connection alive in `snownet`? Is the main reason for keeping it around in this PR fear of breakage?

Reply:
```
2024-06-27T05:31:05.109858Z DEBUG connection{id=5a7e19cc-9928-4123-85bc-9e3d59d01ea6}: boringtun::noise: New session session=794909186
2024-06-27T05:31:15.254256Z DEBUG handle_timeout{id=5a7e19cc-9928-4123-85bc-9e3d59d01ea6}: boringtun::noise::timers: KEEPALIVE(PERSISTENT_KEEPALIVE)
2024-06-27T05:31:26.254579Z DEBUG handle_timeout{id=5a7e19cc-9928-4123-85bc-9e3d59d01ea6}: boringtun::noise::timers: KEEPALIVE(PERSISTENT_KEEPALIVE)
2024-06-27T05:31:27.004141Z  INFO handle_timeout{id=5a7e19cc-9928-4123-85bc-9e3d59d01ea6}: snownet::node: Connection is idle
```

Works as intended. I shortened the timeout to 20s and only sent two pings to the resource after the connection was established. So traffic stopped at 05:31:07. There were two WG keep-alives inbetween, yet the connection got idle closed at :27.